packages/liveblocks-react/src/comments/lib/revalidation.ts

```diff
@@ -0,0 +1,364 @@
+    [manager, revalidateCache]
+  );
+
+  return mutate;
+}
+
+/**
+ * This hook automatically revalidates the cache when the browser comes back online, when the document becomes visible, or periodically.
+ * @param manager - The cache manager
+ * @param revalidateCache - The function used to revalidate the cache
+ * @param options.revalidateOnFocus - If true, the cache will be revalidated when the document becomes visible
+ * @param options.revalidateOnReconnect - If true, the cache will be revalidated when the browser comes back online
+ * @param options.refreshInterval - The interval (in milliseconds) at which the cache should be revalidated
+ */
+export function useAutomaticRevalidation<Data>(
+  manager: CacheManager<Data>,
+  revalidateCache: (shouldDedupe: boolean) => Promise<void>,
+  options: {
+    revalidateOnFocus?: boolean;
+    revalidateOnReconnect?: boolean;
+    refreshInterval?: number;
+  } = {}
+) {
+  const isOnlineRef = useRef(true); // Stores the current online status of the browser
+
+  const {
+    revalidateOnFocus = true,
+    revalidateOnReconnect = true,
+    refreshInterval = 0,
+  } = options;
+
+  /**
+   * Periodically revalidate the cache. The revalidation is skipped if the browser is offline or if the document is not visible.
+   */
+  useEffect(() => {
+    let revalidationTimerId: number;
+
+    function scheduleRevalidation() {
+      if (refreshInterval === 0) return;
+
+      revalidationTimerId = window.setTimeout(() => {
+        const isOnline = isOnlineRef.current;
+        const isVisible = document.visibilityState === "visible";
+
+        // Only revalidate if the browser is online AND document is visible AND there are currently no errors, otherwise schedule the next revalidation
+        if (isOnline && isVisible && !manager.cache?.error) {
+          // Revalidate cache and then schedule the next revalidation
+          void revalidateCache(true).then(scheduleRevalidation);
+          return;
+        }
+
+        scheduleRevalidation();
+      }, refreshInterval);
+    }
+
+    // Schedule the first revalidation
+    scheduleRevalidation();
+
+    return () => {
+      window.clearTimeout(revalidationTimerId);
+    };
+  }, [revalidateCache, refreshInterval, manager]);
+
+  /**
+   * Subscribe to online and offline events to trigger a revalidation when the browser comes back online.
+   * Note: There is a 'navigator.onLine' property that can be used to determine the online status of the browser, but it is not reliable (see https://bugs.chromium.org/p/chromium/issues/detail?id=678075).
+   */
+  useEffect(() => {
+    function handleIsOnline() {
+      isOnlineRef.current = true;
+      const isVisible = document.visibilityState === "visible";
+
+      if (revalidateOnReconnect && isVisible) void revalidateCache(true);
```

Comment:
We perform a revalidation when the browser comes back online (and if the document is visible)